From 2bb95412ed36e97d145d50845abe5c8aba847202 Mon Sep 17 00:00:00 2001
From: Laibin Qiu <qiulaibin@huawei.com>
Date: Mon, 21 Mar 2022 12:25:22 +0800
Subject: [PATCH] blk-mq: add exception handling when srcu->sda alloc failed
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 2bb95412ed36e97d145d50845abe5c8aba847202
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: 186352, https://gitee.com/openeuler/kernel/issues/I4YADX
CVE: NA

--------------------------------

In case of BLK_MQ_F_BLOCKING, per-hctx srcu is used to protect dispatch
critical area. But the current process is not aware when memory of srcu
allocation failed in blk_mq_alloc_hctx, which will leads to illegal
address BUG. Add return value validation to avoid this problem.

Signed-off-by: Laibin Qiu <qiulaibin@huawei.com>
Reviewed-by: Hou Tao <houtao1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 block/blk-mq.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index c3beaca1f4fb..9ae1663348ac 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2852,12 +2852,16 @@ blk_mq_alloc_hctx(struct request_queue *q, struct blk_mq_tag_set *set,
 	if (!hctx->fq)
 		goto free_bitmap;
 
-	if (hctx->flags & BLK_MQ_F_BLOCKING)
-		init_srcu_struct(hctx->srcu);
+	if (hctx->flags & BLK_MQ_F_BLOCKING) {
+		if (init_srcu_struct(hctx->srcu) != 0)
+			goto free_flush_queue;
+	}
 	blk_mq_hctx_kobj_init(hctx);
 
 	return hctx;
 
+ free_flush_queue:
+	blk_free_flush_queue(hctx->fq);
  free_bitmap:
 	sbitmap_free(&hctx->ctx_map);
  free_ctxs:
-- 
2.26.2

